projects/RabbitMQ.Client/client/impl/AsyncConsumerWorkService.cs

```diff
@@ -90,10 +90,7 @@ private async Task Loop()
                         try
                         {
                             Task task = work.Execute();
-                            if (!task.IsCompleted)
```

Comment:
How about doing

```
if (!task.IsCompleted)
{
    await task.ConfigureAwait(false);
}
else
{
    // to materialize exceptions if any
    task.GetAwaiter().GetResult();
}
```

to not undo @stebet optimizations he was trying to achieve here?

Reply:
Btw. I usually prefer to always use the async state machinery to avoid the problems we are seeing here but I totally respect the previous attempts to optimize this code since it is living on the hot path and the optimization was targetting synchronously completing workers

Reply:
The await command should works just the same.
Another solution is check if task.Exception exists, and then throw a new exception.
It will make it more readable.

Reply:
Readability should not be the reason the change the stack trace at all. Plus the task exception is an AggregateException. My proposal properly unpacks the exception should we want to keep the current synchronously completed task optimization